### PR TITLE
Core structure

### PR DESCRIPTION
**Pull request by hypervillain**
Update core structure to prevent dependents to import platform-specific code.

Work in progress: I'm happy with the structure but the changes altered some tests in init that I don't fully understand.
Code in sm-ui has not been migrated. It's pretty anecdotic at the moment. This PR should be shipped with a README in the core that explains what's available, where and (to some extent) why.

### CODE REVIEW

File: packages/init/package.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "lint:precommit": "eslint",
     "audit": "better-npm-audit audit -l high -p",
     "audit-fix": "npm audit fix",
-    "test": "jest"
+    "test": "jest",
+    "slicemachine": "start-slicemachine"
```
Comment (Miniplop) [inline packages/init/package.json]: What this script is doing here ?

File: packages/init/sm.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (Miniplop) [inline packages/init/sm.json]: Why we need a sm.json in the init ?

File: packages/slice-machine/components/App/index.tsx [hunk 3]
```diff
@@ -31,6 +31,7 @@ const RemoveDarkMode: React.FunctionComponent = ({ children }) => {
 type AppProps = {
   theme: () => Theme;
   serverState: ServerState | undefined;
+  children: React.ReactElement;
```
Comment (Miniplop) [inline packages/slice-machine/components/App/index.tsx]: why we need this ? The type React.FunctionComponent already add the children props to the type

File: packages/slice-machine/components/ListItem/index.tsx [hunk 4]
```diff
@@ -13,7 +13,35 @@ import { BsThreeDotsVertical } from "react-icons/bs";
...
+      icon: React.ReactElement;
+    };
+  };
+
+  draggableId: string;
+
+  children: React.ReactElement;
+}
+
+const ListItem = (props: Props) => {
```
Comment (Miniplop) [inline packages/slice-machine/components/ListItem/index.tsx]: We can add React.FunctionComponent type on it

File: packages/slice-machine/lib/builders/CustomTypeBuilder/TabModal/update.tsx [hunk 5]
```diff
@@ -20,7 +20,7 @@ const InputBox = ({
   placeholder: string;
   error?: string;
   [x: string]: any;
-}) => (
+}): React.ReactElement => (
```
Comment (Miniplop) [inline packages/slice-machine/lib/builders/CustomTypeBuilder/TabModal/update.tsx]: It's better to add React.FunctionComponent to the InputBox to type the returned object and props at the same time :)

File: packages/slice-machine/package.json [hunk 6]
```diff
@@ -41,6 +41,7 @@
     "@babel/register": "^7.11.5",
     "@prismicio/slice-canvas-com": "^0.0.4",
     "@reach/menu-button": "^0.16.2",
+    "@segment/analytics-next": "^1.33.0",
```
Comment (Miniplop) [inline packages/slice-machine/package.json:44]: lol that's so scary to see this now, good catch :)

File: packages/slice-machine/src/models/slice/context.tsx [hunk 7]
```diff
@@ -69,15 +69,19 @@ const SliceProvider: React.FunctionComponent<SliceProviderProps> = ({
   );
 };
 
-export const SliceHandler = ({ children }: { children: any }) => {
+export const SliceHandler = ({
```
Comment (Miniplop) [inline packages/slice-machine/src/models/slice/context.tsx]: type the const with React.FunctionComponent, with this you can remove the children type :)

File: packages/core/src/utils/PackageJson.ts [hunk 8]
```diff
@@ -1,21 +0,0 @@
-import * as t from "io-ts";
```
Comment (Miniplop) [inline packages/core/src/utils/PackageJson.ts:1]: I just remember now that Arnaud told me that they use this type in the playground to parse the package.json, I think we should check with the playground team before removing this type completely. 

I still agree that we should remove it and use the one define in the node modules :)

File: packages/slice-machine/tests/apiController/preview.test.ts [hunk 9]
```diff
@@ -13,7 +13,7 @@ jest.mock("@slicemachine/core/build/src/fs-utils", () => ({
 describe("preview controller", () => {
```
Comment (Miniplop) [inline packages/slice-machine/tests/apiController/preview.test.ts:13]: Can you drop this commit ? I already did this here : https://github.com/prismicio/slice-machine/pull/272/commits/1deb9fb4d171bfbbe3998027a6936be2ecb646bd

File: packages/core/src/utils/PackageJson.ts (hunk 8 above)
Comment (hypervillain) [reply packages/core/src/utils/PackageJson.ts:1]: ooooh

Comment (hypervillain) [reply packages/core/src/utils/PackageJson.ts:1]: I moved the type to Models so maybe that'll be enough, I'll ask

File: packages/slice-machine/tests/apiController/preview.test.ts (hunk 9 above)
Comment (hypervillain) [reply packages/slice-machine/tests/apiController/preview.test.ts:13]: I had conflicts in the test and had to resolve manually

File: packages/core/src/utils/PackageJson.ts (hunk 8 above)
Comment (hypervillain) [reply packages/core/src/utils/PackageJson.ts:1]: I found these 2 files that use the core:
- https://github.com/prismicio/playground-libraries/blob/6d1d260fe3d91f83afaebf365f7224d323658755/src/models/Library.ts
- https://github.com/prismicio/playground-libraries/blob/df14d90433c5f253d9db6a8a249642e3f8ce166a/src/MainHandler.ts

but no instance of the PackageJson helper

Comment (hypervillain) [reply packages/core/src/utils/PackageJson.ts:1]: here's the parsing of the package: https://github.com/prismicio/playground-libraries/blob/df14d90433c5f253d9db6a8a249642e3f8ce166a/src/MainHandler.ts#L58

File: packages/core/src/defaults/index.ts [hunk 10]
```diff
@@ -19,4 +19,4 @@ export const DEFAULT_SERVER_PORT = 5555;
 export const DEFAULT_BASE = "https://prismic.io";
 export const USER_SERVICE_BASE = "https://user.internal-prismic.io";
 
-export const ACCEPTED_IMAGE_TYPES = ["png", "jpg", "jpeg"];
```
Comment (MarcMcIntosh) [inline packages/core/src/defaults/index.ts:22]: what was this for? is it to stop people uploading svg's?

File: packages/core/src/internals/files.ts [hunk 11]
```diff
@@ -97,26 +97,26 @@ function readFirstOf<
...
-function mkdir(
+export function mkdir(
   target: string,
   options: { recursive: boolean }
 ): string | undefined {
   return fs.mkdirSync(target, options);
 }
 
-function exists(pathToFile: string): boolean {
+export function exists(pathToFile: string): boolean {
```
Comment (MarcMcIntosh) [inline packages/core/src/internals/files.ts:119]: is this better than `fs.extists`?

Comment (hypervillain) [reply packages/core/src/internals/files.ts:119]: If you need the catch, yes! I'm not sure yet we need this file at all

File: packages/core/README.md [hunk 12]
```diff
@@ -1,3 +1,157 @@
...
+  createManifest(cwd, { apiEndpoint: "https://my-project.prismic.io/api/v2" });
+
+  // my-project
+  const maybeRepoName = maybeRepoNameFromSMFile(cwd, DEFAULT_BASE);
+
+  patchManifest(cwd, { storybook: `https://${maybeRepoName}.vercel.app` });
+
+  const { exists, content } = retrieveManifest(cwd);
+
+  console.log({ exists, content }); // true, { apiEndpoint, storybook }
```
Comment (MarcMcIntosh) [inline packages/core/README.md:77]: console.log

File: packages/core/src/internals/files.ts (hunk 11 above)
Comment (MarcMcIntosh) [reply packages/core/src/internals/files.ts:119]: fs.existsSync() is just true or false, not much else to it.

File: packages/core/README.md (hunk 12 above)
Comment (MarcMcIntosh) [reply packages/core/README.md:77]: I'm pretty sure we can use something like jsdoc to make readme's?

File: packages/core/README.md [hunk 13]
```diff
@@ -1,3 +1,157 @@
...
+                href: string,
+                pathToSlice: string,
+                infos: ComponentInfo,
+                model: SliceAsObject,
+                migrated: boolean,
+            }]
+        ]
+
+    */
+})();
```
Comment (MarcMcIntosh) [inline packages/core/README.md:133]: why have a self invoking function here?

File: packages/core/__tests__/src/fs-utils/framework.test.ts [hunk 14]
```diff
@@ -85,25 +76,9 @@ describe("framework.detectFramework", () => {
...
-      FrameworkUtils.detectFramework(__dirname, Object.values(Frameworks))
-    ).toThrow(wanted);
+    expect(() => detectFramework(__dirname, Object.values(Frameworks))).toThrow(
+      wanted
+    );
     expect(spy).toHaveBeenCalledWith(wanted);
   });
 });
-
-describe("framework.fancyName", () => {
```
Comment (MarcMcIntosh) [inline packages/core/__tests__/src/fs-utils/framework.test.ts:95]: is fancy name still there?

File: packages/core/src/fs-utils/misc.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import path from "path";
+import { Files } from "../internals";
+
+export const findIndexFile = (libPath: string): string | null => {
```
Comment (MarcMcIntosh) [inline packages/core/src/fs-utils/misc.ts:4]: doesn't require/import already do something similar to this?

File: packages/core/src/internals/files.ts [hunk 16]
```diff
@@ -97,26 +97,26 @@ function readFirstOf<
   };
 }
 
-function isDirectory(source: string): boolean {
+export function isDirectory(source: string): boolean {
```
Comment (MarcMcIntosh) [inline packages/core/src/internals/files.ts:100]: by default `lstatSync` will throw if the file is not found. there is the an option to have it return false.
https://nodejs.org/dist/latest-v16.x/docs/api/fs.html#fslstatsyncpath-options

File: packages/core/src/utils/framework.ts [hunk 17]
```diff
@@ -11,52 +9,6 @@ export const UnsupportedFrameWorks = Object.values(Frameworks).filter(
 export const isUnsupported = (framework: Frameworks): boolean =>
   UnsupportedFrameWorks.includes(framework);
 
-function capitaliseFirstLetter(str: string): string {
-  return str.charAt(0).toUpperCase() + str.slice(1);
-}
-
-export function fancyName(str: Frameworks): string {
```
Comment (MarcMcIntosh) [inline packages/core/src/utils/framework.ts:18]: no more fancy name?

File: packages/init/__tests__/maybe-existing-repo.test.ts [hunk 18]
```diff
@@ -14,9 +14,8 @@ import {
 } from "../src/steps/maybe-existing-repo";
 
 import nock from "nock";
-import { Utils } from "@slicemachine/core";
 
-const { Roles } = Utils.roles;
+import { Roles } from "@slicemachine/core/build/src/prismic";
```
Comment (MarcMcIntosh) [inline packages/init/__tests__/maybe-existing-repo.test.ts:18]: prefer not to import from 'core/build/src` I think IDE's do this automatically

File: packages/init/src/steps/configure-project.ts [hunk 19]
```diff
@@ -1,8 +1,38 @@
...
 import { FrameworkResult } from "./detect-framework";
 
-type Base = Utils.Endpoints.Base;
+import {
+  retrieveJsonPackage,
+  patchJsonPackage,
+  patchManifest,
+  retrieveManifest,
+  createManifest,
+} from "@slicemachine/core/build/src/fs-utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/configure-project.ts:10]: 'build' dir?

File: packages/init/src/steps/configure-project.ts [hunk 20]
```diff
@@ -1,8 +1,38 @@
...
+  Base,
+  buildRepositoryEndpoint,
+} from "@slicemachine/core/build/src/prismic";
+import { spinner } from "@slicemachine/core/build/src/internals";
+import {
+  SCRIPT_NAME,
+  SCRIPT_VALUE,
+} from "@slicemachine/core/build/src/defaults";
+import { exists, mkdir } from "@slicemachine/core/build/src/internals/files";
+import { CustomPaths } from "@slicemachine/core/build/src/fs-utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/configure-project.ts:22]: same

File: packages/init/src/steps/create-repo.ts [hunk 21]
```diff
@@ -1,35 +1,36 @@
-import { Communication, Utils } from "@slicemachine/core";
 import type { Models } from "@slicemachine/core";
 
+import { createRepository as createPrismicRepository } from "@slicemachine/core/build/src/prismic";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/create-repo.ts:3]: importing from build dir again

File: packages/init/src/steps/create-repo.ts [hunk 22]
```diff
@@ -1,35 +1,36 @@
...
+import {
+  spinner,
+  bold,
+  writeError,
+} from "@slicemachine/core/build/src/internals";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/create-repo.ts:8]: same

File: packages/init/src/steps/detect-framework.ts [hunk 23]
```diff
@@ -1,16 +1,36 @@
-import { Utils } from "@slicemachine/core";
 import { Models } from "@slicemachine/core";
 import * as inquirer from "inquirer";
 
+import { isUnsupported } from "@slicemachine/core/build/src/utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/detect-framework.ts:4]: same

File: packages/init/src/steps/detect-framework.ts [hunk 24]
```diff
@@ -1,16 +1,36 @@
...
+
 export type FrameworkResult = {
   value: Models.Frameworks;
   manuallyAdded: boolean;
 };
 
+function capitaliseFirstLetter(str: string): string {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+}
+export function fancyName(str: Models.Frameworks): string {
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/detect-framework.ts:21]: fancy name is back :D

File: packages/init/src/steps/display-final-message.ts [hunk 25]
```diff
@@ -1,14 +1,13 @@
-import { Utils, FileSystem } from "@slicemachine/core";
+import { YarnLockPath } from "@slicemachine/core/build/src/fs-utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/display-final-message.ts:1]: build directory, again

File: packages/init/src/steps/install-lib.ts [hunk 26]
```diff
@@ -1,21 +1,17 @@
...
 import path from "path";
 import { getOrElseW } from "fp-ts/Either";
 import { PackageManager, Dependencies } from "../utils/PackageManager";
-import { PackageJsonHelper } from "@slicemachine/core/build/src/utils/PackageJson";
-import {
-  Manifest,
-  ManifestHelper,
-} from "@slicemachine/core/build/src/models/Manifest";
-import Files from "@slicemachine/core/build/src/utils/files";
+import { retrieveJsonPackage } from "@slicemachine/core/build/src/fs-utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/install-lib.ts:9]: same

File: packages/init/src/steps/install-required-dependencies.ts [hunk 27]
```diff
@@ -1,8 +1,18 @@
...
 import { execCommand } from "../utils";
-import { Utils, FileSystem, Models } from "@slicemachine/core";
 
-const {
+import { Models } from "@slicemachine/core";
+import {
+  Files,
+  writeWarning,
+  spinner,
+} from "@slicemachine/core/build/src/internals";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/install-required-dependencies.ts:9]: build directory in import

File: packages/init/src/steps/loginOrBypass.ts [hunk 28]
```diff
@@ -1,10 +1,12 @@
-import { Utils, Auth } from "@slicemachine/core";
-import { UserInfo } from "@slicemachine/core/build/src/core/communication";
+import { Auth } from "@slicemachine/core/build/src/auth";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/loginOrBypass.ts:1]: build dir import

File: packages/init/src/steps/maybe-existing-repo.ts [hunk 29]
```diff
@@ -1,24 +1,37 @@
 import * as inquirer from "inquirer";
 import Separator from "inquirer/lib/objects/separator";
-import { Communication, Utils, FileSystem } from "@slicemachine/core";
-import { Repositories } from "@slicemachine/core/src/models/Repositories";
-import { Repository } from "@slicemachine/core/src/models/Repository";
-import { parsePrismicAuthToken } from "@slicemachine/core/build/src/utils/cookie";
+
+import { Repositories, Repository } from "@slicemachine/core/src/models";
+
+import { DEFAULT_BASE } from "@slicemachine/core/build/src/defaults";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/maybe-existing-repo.ts:6]: build dir again

File: packages/init/src/steps/validate-pkg.ts [hunk 30]
```diff
@@ -1,13 +1,14 @@
-import { FileSystem, Utils } from "@slicemachine/core";
+import { writeError } from "@slicemachine/core/build/src/internals";
+import { retrieveJsonPackage } from "@slicemachine/core/build/src/fs-utils";
```
Comment (MarcMcIntosh) [inline packages/init/src/steps/validate-pkg.ts:2]: build dir again

File: packages/init/src/utils/PackageManager.ts [hunk 31]
```diff
@@ -1,6 +1,8 @@
-import { Utils, FileSystem } from "@slicemachine/core";
 import { execCommand } from "../utils";
 
+import { Files } from "@slicemachine/core/build/src/internals";
```
Comment (MarcMcIntosh) [inline packages/init/src/utils/PackageManager.ts:3]: build dir again

File: packages/slice-machine/changelog/migrate.js [hunk 32]
```diff
@@ -1,10 +1,10 @@
 const path = require("path");
```
Comment (MarcMcIntosh) [inline packages/slice-machine/changelog/migrate.js:1]: you know there's tools out there to get change-logs from git commits?

File: packages/slice-machine/components/Grid/index.tsx [hunk 33]
```diff
@@ -7,8 +7,8 @@ const Grid = ({
 }: {
   elems: any;
   gridTemplateMinPx?: string;
-  renderElem: Function;
-}) => {
+  renderElem: (elem: any, i: number) => React.ReactElement | null;
+}): React.ReactElement => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/components/Grid/index.tsx:11]: does `React.ReactNode` also support null?

File: packages/slice-machine/lib/builders/SliceBuilder/Header/VariationModal.tsx [hunk 34]
```diff
@@ -16,7 +16,7 @@ const Error = ({ msg }: { msg?: string }) => (
 );
 const VariationModal: React.FunctionComponent<{
   isOpen: boolean;
-  onClose: () => any;
+  onClose: () => unknown;
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/builders/SliceBuilder/Header/VariationModal.tsx:19]: not void?

File: packages/slice-machine/lib/models/common/CustomType/index.ts [hunk 35]
```diff
@@ -55,6 +55,7 @@ export const CustomType = {
     };
   },
   toJsonModel(ct: CustomType<ObjectTabs>): CustomTypeJsonModel {
+    // eslint no-unused-vars: ["error", { "ignoreRestSiblings": true }]
     const { tabs, previewUrl, ...rest } = ct;
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/models/common/CustomType/index.ts:59]: ct.tabs is called further down in this file

File: packages/slice-machine/package.json [hunk 36]
```diff
@@ -29,9 +29,9 @@
     "export-build": "next export && npm run clean-build",
     "bundle": "npm run build && npm run test && npm run export-build",
     "slicemachine": "start-slicemachine --port 9999",
-    "lint": "eslint --max-warnings 1067 --ext .ts,.tsx .",
+    "lint": "eslint --max-warnings 983 --ext .ts,.tsx .",
```
Comment (MarcMcIntosh) [inline packages/slice-machine/package.json:32]: nice

File: packages/slice-machine/server/src/api/auth/start.ts [hunk 37]
```diff
@@ -7,13 +8,13 @@ type StartResponse = {
...
-    FileSystem.PrismicSharedConfigManager.setProperties({
-      cookies: Utils.Cookie.serializeCookies([]),
+    SharedConfigManager.setProperties({
+      cookies: serializeCookies([]),
     });
     return {};
   } catch (e) {
     return {
-      err: e,
+      err: e as Error,
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/auth/start.ts:17]: does `catch (e: Error)` not work?

File: packages/slice-machine/server/src/api/common/hooks/onSaveSlice.ts [hunk 38]
```diff
@@ -37,10 +38,8 @@ const createIndexFileForFrameWork = (
   return createIndexFile(lib);
 };
 
-export default async function onSaveSlice(
-  env: BackendEnvironment
-): Promise<void> {
-  const libraries = await Libraries.libraries(
+export default function onSaveSlice(env: BackendEnvironment): void {
+  const libraries = getLibraries(
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/common/hooks/onSaveSlice.ts:42]: no longer async?

File: packages/slice-machine/server/src/api/libraries.ts [hunk 39]
```diff
@@ -1,4 +1,5 @@
-import { Libraries, Models } from "@slicemachine/core";
+import { Models } from "@slicemachine/core";
+import { libraries as getLibraries } from "@slicemachine/core/build/src/libraries";
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/libraries.ts:2]: should we not rename `libraries` to `getLibraries`?

File: packages/slice-machine/src/models/customType/actions/push.ts [hunk 40]
```diff
@@ -6,8 +6,8 @@ import { ToastPayload } from "@src/ToastProvider/utils";
 export default function push(
   dispatch: ({ type, payload }: { type: string; payload?: any }) => void
 ) {
-  return async (ct: CustomTypeState, setData: (data: ToastPayload) => void) => {
-    await fetchApi({
+  return (ct: CustomTypeState, setData: (data: ToastPayload) => void): void => {
+    fetchApi({
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/models/customType/actions/push.ts:10]: how are errors handled here?

File: packages/slice-machine/src/models/slice/actions/save.ts [hunk 41]
```diff
@@ -8,10 +8,7 @@ import { SliceSaveResponse } from "@lib/models/common/Slice";
 export default function save(
   dispatch: ({ type, payload }: { type: string; payload?: any }) => void
 ) {
-  return async (
-    slice: SliceState,
-    setData: (data: ToastPayload) => void = () => {}
-  ) => {
+  return (slice: SliceState, setData?: (data: ToastPayload) => void): void => {
     fetchApi({
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/models/slice/actions/save.ts:12]: why note return Promise<void> so the function can be marked as async?

File: packages/slice-machine/src/models/slice/context.tsx [hunk 42]
```diff
@@ -69,15 +69,15 @@ const SliceProvider: React.FunctionComponent<SliceProviderProps> = ({
   );
 };
 
-export const SliceHandler = ({ children }: { children: any }) => {
+export const SliceHandler: React.FunctionComponent = ({ children }) => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/models/slice/context.tsx:72]: React.Node would work as a return type here :)

File: packages/slice-machine/src/models/slice/store.ts [hunk 43]
```diff
@@ -43,7 +43,7 @@ export default class SliceStore implements Store {
 
   variation = (variationId: string): Record<string, Function> => {
     return {
-      generateScreenShot: generateScreenShot(this.dispatch)(variationId),
+      generateScreenShot: generateScreenShot(this.dispatch)(),
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/models/slice/store.ts:46]: no more variation id?

File: packages/slice-machine/src/models/slice/variation/actions/screenshot.ts [hunk 44]
```diff
@@ -4,10 +4,20 @@ import { ActionType } from "./ActionType";
...
+    type,
+    payload,
+  }: {
+    type: string;
+    payload?: { screenshots: Record<string, ScreenshotUI> };
+  }) => void
 ) {
-  return (_variationId: string) => {
-    return async (
+  return (): ((
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/models/slice/variation/actions/screenshot.ts:15]: this is async right?

**Review by MarcMcIntosh — CHANGES_REQUESTED** [review]
Good work, I have a few questions

**Comment by Miniplop** [comment]
As this [PR](https://github.com/prismicio/slice-machine/pull/411) is now merged this PR is deprecated

**Pull request closed without merging**